### PR TITLE
Make op unique name generation case insensitive

### LINKED ISSUE #16261: Tensorflow Debug tfdbg ValueError with combined loss functions.
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes. Included
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10 Home x64, 1709
- **TensorFlow installed from (source or binary)**: binary (pip3)
- **TensorFlow version (use command below)**:  1.4.0
- **Python version**:  3.5.2
- **Exact command to reproduce**: See source code.

### Describe the problem
When attempting to combine two loss functions, tfdbg fails to properly grab the gradients due to a ValueError when executing run.

### Source code / logs
[consolelog.txt](https://github.com/tensorflow/tensorflow/files/1649424/consolelog.txt) Log of the error.
[not_working.txt](https://github.com/tensorflow/tensorflow/files/1649426/not_working.txt) Code example that generates the error.
[working.txt](https://github.com/tensorflow/tensorflow/files/1649427/working.txt) Code example where the two loss functions are split that does not generate the error.

### PR DESCRIPTION
**Pull request by pvaneck**
Unique name generation for operations depends on checking a dict for names currently in use. This commit makes it so that the names stored in this dict are always lowercase so that we can check if a name already exists regardless of the capitalization.

This helps in filesystems where file paths are case insensitive and tensor dumps (like with tfdbg) try to follow directory structures that correspond to the tensor names. If two tensors have names with the same spelling, but different capitalizations, then this can lead to unintended side effects/errors on these case-insensitive file systems.

These are likely just edge cases, but one such issue where this can be a problem is shown in #16261. Here, because both tf.multiply and the '\*' overloaded operator were used, this caused name collisions for some intermediate tensors where you would have  tensor names like "gradients/mul_grad/tuple/control_dependency_1" and "gradients/Mul_grad/Mul" (since tf.multiply uses name="Mul" while using the '\*' operator uses name="mul"). So when these tensors are dumped, only one "mul_grad" or "Mul_grad" directory can be created on specific file systems (like with Mac and Windows).

Fixes #16261

### CODE REVIEW

**Comment by jhseu** [comment]
This seems reasonable, but adding @asimshankar to review since this changes the names of ops that have this collision.

**Comment by asimshankar** [comment]
@skye might have a more informed opinion, reassigning :)

File: tensorflow/python/framework/ops.py [hunk 1]
```diff
@@ -3430,8 +3430,9 @@ def _create_op_from_tf_operation(self, c_op, compute_device=True):
     # the name will still appear in _names_in_use even though the name hasn't
     # been used. This is ok, just leave _names_in_use as-is in this case.
     # TODO(skyewm): make the C API guarantee no name conflicts.
-    if ret.name not in self._names_in_use:
-      self._names_in_use[ret.name] = 1
+    name = ret.name.lower()
```
Comment (skye) [inline tensorflow/python/framework/ops.py]: nit: maybe call this name_key to be consistent with unique_name and make it clear that we're not changing the actual name.

**Review by skye — DISMISSED** [review]
Seems fine to me. Thanks for the nice PR.

**Comment by pvaneck** [comment]
Some tests failing due to name changes. Will investigate.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jhseu: It has been 19 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by caisq** [comment]
@skye can you give a formal approval so we can merge this PR?

**Review by skye — APPROVED** [review]

**Comment by eladeban** [comment]
I think its a bad idea.
Old checkpoints will not load properly, this has the potential to break a bunch of saved models on all file systems, just to make file-systems with non case sensitive naming be more friendly with a particular use case.
